### PR TITLE
feat (provider/google-vertex): Add imagen support.

### PR DESCRIPTION
**Pull request by shaper**
Part of #4088

### CODE REVIEW

File: packages/google-vertex/src/google-vertex-image-model.ts [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+  private getAspectRatio(width: number, height: number): string {
+    // Map common dimensions to Imagen's supported aspect ratios
+    if (width === height) return '1:1';
+    if (width === 896 && height === 1280) return '3:4';
+    if (width === 1280 && height === 896) return '4:3';
+    if (width === 768 && height === 1408) return '9:16';
+    if (width === 1408 && height === 768) return '16:9';
+
+    // Default to 1:1 if no match
+    return '1:1';
```
Comment (lgrammel) [inline packages/google-vertex/src/google-vertex-image-model.ts:82]: can we do some clever math here instead of relying on fixed values (ask llm)? or are these related to the imagen output sizes?

File: examples/ai-core/src/e2e/google-vertex.test.ts [hunk 2]
```diff
@@ -26,6 +27,7 @@ const MODEL_VARIANTS = {
     // 'gemini-1.0-pro-001',
   ],
   embedding: ['textembedding-gecko', 'textembedding-gecko-multilingual'],
+  image: ['imagen-3.0-generate-001', 'imagen-3.0-fast-generate-001'],
```
Comment (lgrammel) [inline examples/ai-core/src/e2e/google-vertex.test.ts:30]: is it possible to run specific e2e tests easily? as in all image tests or a specific image test?

File: examples/ai-core/src/e2e/google-vertex.test.ts [hunk 3]
```diff
@@ -453,5 +455,38 @@ describe.each(Object.values(RUNTIME_VARIANTS))(
...
+
+        // Verify we got a Uint8Array back
+        expect(image.uint8Array).toBeInstanceOf(Uint8Array);
+
+        // Check the file size is reasonable (at least 10KB, less than 10MB)
+        expect(image.uint8Array.length).toBeGreaterThan(10 * 1024);
+        expect(image.uint8Array.length).toBeLessThan(10 * 1024 * 1024);
+
+        // Verify PNG format by checking magic numbers
+        const pngSignature = [137, 80, 78, 71, 13, 10, 26, 10];
```
Comment (lgrammel) [inline examples/ai-core/src/e2e/google-vertex.test.ts:476]: we have a mime type detection helper for this but I'm not sure it's exposed

File: packages/google-vertex/src/google-vertex-image-model.ts (hunk 1 above)
Comment (lgrammel) [reply packages/google-vertex/src/google-vertex-image-model.ts:82]: also, is the aspect ratio required? otherwise we could rely on provider options so users can specify it

File: packages/google-vertex/src/google-vertex-image-model.ts [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+    prompt: string;
+    n: number;
+    size: `${number}x${number}` | undefined;
+    providerOptions: Record<string, Record<string, JSONValue>>;
+    abortSignal?: AbortSignal;
+    headers?: Record<string, string>;
+  }): Promise<{ images: string[] }> {
+    const [width, height] = (options.size ?? '1024x1024')
+      .split('x')
+      .map(Number);
```
Comment (lgrammel) [inline packages/google-vertex/src/google-vertex-image-model.ts:38]: are these only needed for aspect ratio?

File: packages/google-vertex/src/google-vertex-image-model.ts [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+
+    const response = await (this.options.fetch ?? fetch)(
```
Comment (lgrammel) [inline packages/google-vertex/src/google-vertex-image-model.ts:40]: should use postJsonToApi similar to `openai-image-model`

File: packages/google-vertex/src/google-vertex-image-model.ts [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+            sampleCount: options.n,
+            aspectRatio: this.getAspectRatio(width, height),
+            ...options.providerOptions,
+          },
+        }),
+      },
+    );
+
+    if (!response.ok) {
+      throw new Error(`HTTP error! status: ${response.status}`);
```
Comment (lgrammel) [inline packages/google-vertex/src/google-vertex-image-model.ts:62]: error parsing?

File: packages/google-vertex/src/google-vertex-image-model.ts [hunk 7]
```diff
@@ -0,0 +1,84 @@
...
+    private options: GoogleVertexImageModelOptions,
+  ) {}
+
+  async doGenerate(options: {
+    prompt: string;
+    n: number;
+    size: `${number}x${number}` | undefined;
+    providerOptions: Record<string, Record<string, JSONValue>>;
+    abortSignal?: AbortSignal;
+    headers?: Record<string, string>;
```
Comment (lgrammel) [inline packages/google-vertex/src/google-vertex-image-model.ts:34]:
```
  }: Parameters<ImageModelV1['doGenerate']>[0]): Promise<
    Awaited<ReturnType<ImageModelV1['doGenerate']>>
 ```

File: packages/google-vertex/src/google-vertex-image-model.ts [hunk 8]
```diff
@@ -0,0 +1,84 @@
...
+    if (!response.ok) {
+      throw new Error(`HTTP error! status: ${response.status}`);
+    }
+
+    const data = await response.json();
+    return {
+      images: data.predictions.map(
+        (p: { bytesBase64Encoded: string }) => p.bytesBase64Encoded,
+      ),
+    };
```
Comment (lgrammel) [inline packages/google-vertex/src/google-vertex-image-model.ts:70]: What about output validation?

File: packages/google-vertex/src/google-vertex-image-model.ts [hunk 9]
```diff
@@ -0,0 +1,84 @@
...
+export type GoogleVertexImageModelId =
+  | 'imagen-3.0-generate-001'
+  | 'imagen-3.0-fast-generate-001';
+
+interface GoogleVertexImageModelOptions {
+  provider: string;
+  baseURL: string;
+  headers?: Resolvable<Record<string, string | undefined>>;
+  fetch?: typeof fetch;
+}
```
Comment (lgrammel) [inline packages/google-vertex/src/google-vertex-image-model.ts:13]: wonder if we can extract generic model options for google (across different kinds of models)

File: packages/google-vertex/src/google-vertex-image-model.ts (hunk 1 above)
Comment (shaper) [reply packages/google-vertex/src/google-vertex-image-model.ts:82]: https://cloud.google.com/vertex-ai/generative-ai/docs/image/generate-images#aspect-ratio
https://cloud.google.com/vertex-ai/generative-ai/docs/image/generate-images#performance-limits

It's an optional parameter and the default is 1:1. It appears the only way to specify the desired image size is via specifying an aspect ratio, which in turn maps to an output resolution width/height per the above links.

Changing it to use a provider option and to throw if a user tries to specify an explicit size. We could allow size if it's one of the supported sizes, but that's also more maintenance work for us.

File: examples/ai-core/src/e2e/google-vertex.test.ts (hunk 2 above)
Comment (shaper) [reply examples/ai-core/src/e2e/google-vertex.test.ts:30]: Yes, you can use `-t <filter string>` so `-t image` to match `it` or `describe` strings. You can use `-t 'Chat Model'` or `-t 'Image Model'` for example to run all test cases under the respective `describe` subset. I added a note about this in the [README](https://github.com/vercel/ai/tree/main/examples/ai-core#end-to-end-provider-integration-tests) although I didn't fully elaborate as much as above.

File: packages/google-vertex/src/google-vertex-image-model.ts (hunk 4 above)
Comment (shaper) [reply packages/google-vertex/src/google-vertex-image-model.ts:38]: Yes, removed now, the LLM was basically reverse engineering the aspect ratio from user specified size, and defaulting to the 1:1 imagen defaults to. That's one way to do it, but it's odd and could lead to user surprise since they might not get the size they specified.

File: examples/ai-core/src/e2e/google-vertex.test.ts (hunk 2 above)
Comment (shaper) [reply examples/ai-core/src/e2e/google-vertex.test.ts:30]: I don't know if you can use `-t` in some manner to specify you only want to run tests for a particular model, though. I have been commenting/uncommenting them in the list when I want to do that.

File: examples/ai-core/src/e2e/google-vertex.test.ts (hunk 3 above)
Comment (shaper) [reply examples/ai-core/src/e2e/google-vertex.test.ts:476]: Exposed it for use, much nicer.

File: packages/google-vertex/src/google-vertex-image-model.ts (hunk 9 above)
Comment (shaper) [reply packages/google-vertex/src/google-vertex-image-model.ts:13]: I think we can and it's a good idea but would like to do it as a separate change -- will note for follow-up.

File: packages/google-vertex/src/google-vertex-image-model.ts (hunk 5 above)
Comment (shaper) [reply packages/google-vertex/src/google-vertex-image-model.ts:40]: Done.

File: packages/google-vertex/src/google-vertex-image-model.ts (hunk 7 above)
Comment (shaper) [reply packages/google-vertex/src/google-vertex-image-model.ts:34]: Done.

File: packages/google-vertex/src/google-vertex-image-model.ts (hunk 6 above)
Comment (shaper) [reply packages/google-vertex/src/google-vertex-image-model.ts:62]: Reworked as part of postToJsonApi edit.

File: packages/google-vertex/src/google-vertex-image-model.ts (hunk 8 above)
Comment (shaper) [reply packages/google-vertex/src/google-vertex-image-model.ts:70]: Fixed via postToJsonApi rework.

File: packages/ai/core/index.ts [hunk 10]
```diff
@@ -11,3 +11,4 @@ export * from './registry';
 export * from './tool';
 export * from './types';
 export { cosineSimilarity } from './util/cosine-similarity';
+export { detectImageMimeType } from './util/detect-image-mimetype';
```
Comment (lgrammel) [inline packages/ai/core/index.ts:14]: Torn about exposing the utility. We'd need to document / support it. We could expose it under `/test` or `/internal`?

**Review by lgrammel — APPROVED** [review]

File: packages/ai/core/index.ts (hunk 10 above)
Comment (lgrammel) [reply packages/ai/core/index.ts:14]: or even copy the utility for now until we can think more about how to best do it. I'm leaning towards including under `ai/test`

Comment (shaper) [reply packages/ai/core/index.ts:14]: Will copy it for now.
